File: app/src/main/java/at/bitfire/davdroid/log/LogcatHandler.java
/*
 * Copyright © 2013 – 2016 Ricki Hirner (bitfire web engineering).
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Public License v3.0
 * which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/gpl.html
 */

package at.bitfire.davdroid.log;

import android.util.Log;

import org.apache.commons.lang3.math.NumberUtils;

import java.util.logging.Handler;
import java.util.logging.Level;
import java.util.logging.LogRecord;

public class LogcatHandler extends Handler {
    private static final int MAX_LINE_LENGTH = 3000;
    public static final LogcatHandler INSTANCE = new LogcatHandler();

    private LogcatHandler() {
        super();
        setFormatter(PlainTextFormatter.LOGCAT);
        setLevel(Level.ALL);
    }

    @Override
    public void publish(LogRecord r) {
        String text = getFormatter().format(r);
        int level = r.getLevel().intValue();

        int end = text.length();
        for (int pos = 0; pos < end; pos += MAX_LINE_LENGTH) {
            String line = text.substring(pos, NumberUtils.min(pos + MAX_LINE_LENGTH, end));

            if (level >= Level.SEVERE.intValue())
                Log.e(r.getLoggerName(), line);
            else if (level >= Level.WARNING.intValue())
                Log.w(r.getLoggerName(), line);
            else if (level >= Level.CONFIG.intValue())
                Log.i(r.getLoggerName(), line);
            else if (level >= Level.FINER.intValue())
                Log.d(r.getLoggerName(), line);
            else
                Log.v(r.getLoggerName(), line);
        }
    }

    @Override
    public void flush() {
    }

    @Override
    public void close() {
    }

}

File: app/src/main/java/at/bitfire/davdroid/log/LogcatHandler.kt
/*
 * Copyright © Ricki Hirner (bitfire web engineering).
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Public License v3.0
 * which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/gpl.html
 */

package at.bitfire.davdroid.log

import android.util.Log

import org.apache.commons.lang3.math.NumberUtils

import java.util.logging.Handler
import java.util.logging.Level
import java.util.logging.LogRecord

object LogcatHandler: Handler() {

    private val MAX_LINE_LENGTH = 3000

    init {
        formatter = PlainTextFormatter.LOGCAT
        level = Level.ALL
    }

    override fun publish(r: LogRecord) {
        val text = formatter.format(r)
        val level = r.level.intValue()

        val end = text.length
        var pos = 0
        while (pos < end) {
            val line = text.substring(pos, NumberUtils.min(pos + MAX_LINE_LENGTH, end))
            when {
                level >= Level.SEVERE.intValue()  -> Log.e(r.loggerName, line)
                level >= Level.WARNING.intValue() -> Log.w(r.loggerName, line)
                level >= Level.CONFIG.intValue()  -> Log.i(r.loggerName, line)
                level >= Level.FINER.intValue()   -> Log.d(r.loggerName, line)
                else                              -> Log.v(r.loggerName, line)
            }
            pos += MAX_LINE_LENGTH
        }
    }

    override fun flush() {}
    override fun close() {}

}
